File: tauri-app/src-tauri/src/providers/zai.rs
use crate::providers::models::*;
use crate::providers::Provider;
use std::future::Future;
use std::pin::Pin;

pub struct ZaiProvider {
    client: reqwest::Client,
    api_key: String,
    region: String,
}

impl ZaiProvider {
    pub fn new(client: reqwest::Client, api_key: &str, region: Option<&str>) -> Self {
        Self {
            client,
            api_key: api_key.to_string(),
            region: region.unwrap_or("china").to_string(),
        }
    }

    fn base_url(&self) -> String {
        match self.region.as_str() {
            "china" => "https://open.bigmodel.cn".to_string(),
            "global" => "https://api.z.ai".to_string(),
            other if other.starts_with("http") => other.to_string(),
            other => format!("https://{}", other),
        }
    }
}

impl Provider for ZaiProvider {
    fn fetch_usage(&self) -> Pin<Box<dyn Future<Output = UsageData> + Send + '_>> {
        Box::pin(async move {
        if self.api_key.is_empty() {
            return UsageData::no_api_key("zai", "Z.ai (智谱)");
        }

        let url = format!("{}/api/monitor/usage/quota/limit", self.base_url());

        let resp = match self.client
            .get(&url)
            .header("Authorization", format!("Bearer {}", self.api_key))
            .header("Accept", "application/json")
            .timeout(std::time::Duration::from_secs(15))
            .send()
            .await
        {
            Ok(r) => r,
            Err(e) => return UsageData::error("zai", "Z.ai (智谱)", &e.to_string()),
        };

        if resp.status() == reqwest::StatusCode::UNAUTHORIZED {
            return UsageData::unauthorized("zai", "Z.ai (智谱)", "API 密钥无效或已过期");
        }

        let data: serde_json::Value = match resp.json().await {
            Ok(d) => d,
            Err(e) => return UsageData::error("zai", "Z.ai (智谱)", &e.to_string()),
        };

        let code = data.get("code").and_then(|v| v.as_i64()).unwrap_or(-1);
        if code != 200 {
            let msg = data
                .get("msg")
                .and_then(|v| v.as_str())
                .unwrap_or("未知错误");
            return UsageData::error("zai", "Z.ai (智谱)", &format!("API 错误: {}", msg));
        }

        let payload = data.get("data").cloned().unwrap_or(serde_json::json!({}));
        let plan_name = payload
            .get("planName")
            .or(payload.get("plan"))
            .or(payload.get("packageName"))
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();

        let mut windows: Vec<UsageWindow> = Vec::new();

        if let Some(limits) = payload.get("limits").and_then(|v| v.as_array()) {
            for limit in limits {
                let limit_type = limit
                    .get("type")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .to_string();
                let usage = limit.get("usage").and_then(|v| v.as_f64()).unwrap_or(0.0);
                let number = limit.get("number").and_then(|v| v.as_f64()).unwrap_or(0.0);
                let remaining = limit
                    .get("remaining")
                    .and_then(|v| v.as_f64())
                    .unwrap_or(0.0);
                let current_value = limit
                    .get("currentValue")
                    .and_then(|v| v.as_f64())
                    .unwrap_or(usage);
                let percentage = limit.get("percentage").and_then(|v| v.as_f64());

                let pct = percentage
                    .unwrap_or_else(|| if number > 0.0 { (current_value / number) * 100.0 } else { 0.0 });

                let resets_at = limit
                    .get("nextResetTime")
                    .and_then(|v| v.as_i64())
                    .map(|ms| {
                        chrono::DateTime::from_timestamp_millis(ms)
                            .map(|dt| dt.format("%Y-%m-%dT%H:%M:%S%:z").to_string())
                            .unwrap_or_default()
                    });

                let (label, unit) = if limit_type == "TIME_LIMIT" {
                    ("MCP/时间配额".to_string(), "次".to_string())
                } else {
                    ("Token 配额".to_string(), "tokens".to_string())
                };

                windows.push(UsageWindow {
                    label,
                    used_percent: pct.min(100.0),
                    used: current_value,
                    total: number,
                    remaining,
                    resets_at,
                    unit,
                });
            }
        }

        // Token 配额排在前面
        windows.sort_by(|a, b| {
            let a_has_token = a.label.contains("Token") as i32;
            let b_has_token = b.label.contains("Token") as i32;
            b_has_token.cmp(&a_has_token)
        });

        let is_empty = windows.is_empty();
        UsageData {
            provider_id: "zai".into(),
            provider_name: "Z.ai (智谱)".into(),
            status: if is_empty {
                ProviderStatus::Error
            } else {
                ProviderStatus::Ok
            },
            plan_name,
            windows,
            balance: None,
            error_message: if is_empty {
                "未找到配额数据".into()
            } else {
                String::new()
            },
            updated_at: UsageData::now_timestamp(),
        }
        })
    }
}

File: src/providers/zai.py
"""Z.ai (Zhipu/BigModel) provider - China domestic version."""

from __future__ import annotations

import os
from datetime import datetime, timezone

from .base import BaseProvider, ProviderStatus, UsageData, UsageWindow


class ZaiProvider(BaseProvider):
    """Z.ai (Zhipu BigModel) coding plan usage provider.

    Supports both global (api.z.ai) and China (open.bigmodel.cn) endpoints.
    """

    DEFAULT_HOSTS = {
        "china": "https://open.bigmodel.cn",
        "global": "https://api.z.ai",
    }

    def __init__(self, api_key: str = "", region: str = "china", **kwargs):
        super().__init__(
            provider_id="zai",
            name="Z.ai (智谱)",
            description="智谱 BigModel coding plan usage",
            requires_api_key=True,
            env_key="Z_AI_API_KEY",
            api_key=api_key,
            extra_config={"region": region},
            **kwargs,
        )
        self.region = region

    @property
    def base_url(self) -> str:
        host = self.extra_config.get("region", self.region)
        if host in self.DEFAULT_HOSTS:
            host = self.DEFAULT_HOSTS[host]
        if not host.startswith("http"):
            host = f"https://{host}"
        custom_url = os.environ.get("Z_AI_QUOTA_URL")
        if custom_url:
            return custom_url
        return host

    async def fetch_usage(self) -> UsageData:
        token = self.get_api_key()
        if not token:
            return self._no_key_result()

        url = f"{self.base_url}/api/monitor/usage/quota/limit"
        headers = {
            "Authorization": f"Bearer {token}",
            "Accept": "application/json",
        }

        result = await self._http_get(url, headers)
        if isinstance(result, UsageData):
            return result
        return self._parse_response(result)

    def _parse_response(self, data: dict) -> UsageData:
        code = data.get("code", -1)
        if code != 200:
            return self._error_result(data.get("msg", f"API 错误码: {code}"))

        payload = data.get("data", {})
        plan_name = (
            payload.get("planName")
            or payload.get("plan")
            or payload.get("packageName")
            or ""
        )

        windows: list[UsageWindow] = []
        for limit in payload.get("limits", []):
            limit_type = limit.get("type", "")
            usage = limit.get("usage", 0)
            number = limit.get("number", 0)
            remaining = limit.get("remaining", 0)
            current_value = limit.get("currentValue", usage)
            percentage = limit.get("percentage")

            unit_num = limit.get("unit", 1)
            # unit represents duration: 1=minute, 5=minutes, 60=hour etc.
            # The number field is the window size
            window_size = number

            resets_at = None
            reset_ms = limit.get("nextResetTime")
            if reset_ms:
                try:
                    resets_at = datetime.fromtimestamp(reset_ms / 1000, tz=timezone.utc)
                except (OSError, ValueError):
                    pass

            if percentage is None and window_size > 0:
                percentage = (current_value / window_size) * 100

            label = "Token 配额"
            if limit_type == "TIME_LIMIT":
                label = "MCP/时间配额"
            elif limit_type == "TOKENS_LIMIT":
                label = "Token 配额"

            unit = "次" if limit_type == "TIME_LIMIT" else "tokens"

            windows.append(
                UsageWindow(
                    label=label,
                    used_percent=min(percentage or 0, 100),
                    used=float(current_value),
                    total=float(window_size),
                    remaining=float(remaining),
                    resets_at=resets_at,
                    unit=unit,
                )
            )

        # Token 配额排在前面，MCP/时间配额排在后面
        windows.sort(key=lambda w: 0 if "Token" in w.label else 1)

        return UsageData(
            provider_id=self.provider_id,
            provider_name=self.name,
            status=ProviderStatus.OK,
            plan_name=plan_name,
            windows=windows,
            raw_response=data,
        )

    def get_display_config(self) -> dict[str, str]:
        return {
            "api_key": "password",
            "region": "select:china,global",
        }
